File: arc/runtime/telem/source.go
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

package telem

import (
	"context"

	"github.com/synnaxlabs/arc/ir"
	"github.com/synnaxlabs/arc/runtime/node"
	"github.com/synnaxlabs/arc/runtime/state"
	"github.com/synnaxlabs/x/query"
	xtelem "github.com/synnaxlabs/x/telem"
)

var (
	sourceSymbolName = "on"
	sourceSymbol     = ir.Symbol{
		Name: sourceSymbolName,
		Kind: ir.KindStage,
		Type: ir.Stage{
			Config: ir.NamedTypes{
				Keys:   []string{"channel"},
				Values: []ir.Type{ir.Chan{ValueType: ir.NewTypeVariable("T", nil)}},
			},
			Outputs: ir.NamedTypes{
				Keys:   []string{ir.DefaultOutputParam},
				Values: []ir.Type{ir.NewTypeVariable("T", nil)},
			},
		},
	}
	SymbolResolver = ir.MapResolver{sourceSymbolName: sourceSymbol}
)

type source struct {
	node          ir.Node
	telem         *State
	state         *state.State
	key           uint32
	highWaterMark xtelem.Alignment
}

func (s *source) Init(context.Context, func(output string)) {}

func (s *source) Next(_ context.Context, onOutputChange func(param string)) {
	for _, ser := range s.telem.Data[s.key].Series {
		ab := ser.AlignmentBounds()
		if ab.Upper > s.highWaterMark {
			s.highWaterMark = ab.Upper
			s.state.Outputs[ir.Handle{Param: ir.DefaultOutputParam, Node: s.node.Key}] = ser
			onOutputChange(ir.DefaultOutputParam)
		}
	}
}

type telemFactory struct {
	telem *State
}

func (t telemFactory) Create(_ context.Context, cfg node.Config) (node.Node, error) {
	if cfg.Node.Type != sourceSymbolName {
		return nil, query.NotFound
	}
	key := cfg.Node.Channels.Read.Keys()[0]
	t.telem.register(key, cfg.Node.Key)
	return &source{
		node:          cfg.Node,
		telem:         t.telem,
		state:         cfg.State,
		key:           key,
		highWaterMark: 0,
	}, nil
}

func NewTelemFactory(state *State) node.Factory {
	return &telemFactory{telem: state}
}

File: pluto/src/telem/aether/test/source.ts
// Copyright 2025 Synnax Labs, Inc.
//
// Use of this software is governed by the Business Source License included in the file
// licenses/BSL.txt.
//
// As of the Change Date specified in that file, in accordance with the Business Source
// License, use of this software will be governed by the Apache License, Version 2.0,
// included in the file licenses/APL.txt.

import { type destructor, id, observe } from "@synnaxlabs/x";

import {
  type BooleanSourceSpec,
  type NumberSourceSpec,
  type Source,
  type StringSourceSpec,
  type Telem,
} from "@/telem/aether/telem";
import { registerInstance, TEST_SOURCE_TYPE } from "@/telem/aether/test/factory";

export class TestSource<V> extends observe.Observer<void> implements Source<V>, Telem {
  static readonly TYPE = TEST_SOURCE_TYPE;
  readonly id: string;
  private _value: V;
  private _destructor: destructor.Destructor;

  constructor(initialValue: V) {
    super();
    this.id = id.create();
    this._value = initialValue;
    this._destructor = registerInstance(this.id, this);
  }

  value(): V {
    return this._value;
  }

  setValue(v: V): void {
    this._value = v;
    this.notify();
  }

  cleanup(): void {
    this._destructor();
  }
}

export const source = <V>(initialValue: V): TestSource<V> =>
  new TestSource(initialValue);

export const booleanSourceSpec = (source: TestSource<boolean>): BooleanSourceSpec => ({
  type: TestSource.TYPE,
  props: { testId: source.id },
  variant: "source",
  valueType: "boolean",
});

export const numberSourceSpec = (source: TestSource<number>): NumberSourceSpec => ({
  type: TestSource.TYPE,
  props: { testId: source.id },
  variant: "source",
  valueType: "number",
});

export const stringSourceSpec = (source: TestSource<string>): StringSourceSpec => ({
  type: TestSource.TYPE,
  props: { testId: source.id },
  variant: "source",
  valueType: "string",
});
